### ISSUE TITLE
`NumberFormatter` & co. scope (monolithic vs modular)

### ISSUE THREAD

**Issue by zbraniecki** [unicode-org/icu4x#275]
In ICU (and ECMA402) `NumberFormat` becomes the jack of all trades with formatting for numbers, currencies, measuring units, and so on. There's even a drive from Shane to incorporate Pluralization as a feature of a `NumberFormat`.

Shane justified it by saying that all number formatters take similar options and similar operations to tailor the data.

The cost of such approach is that it becomes trickier to modularize such crates and `NumberFormat` becomes actually a pretty large codebase on a very fundamental level that is required by basically everything and fragile DCE is the only hope to keep the overhead lower.

It is my impression that in ICU4X context we can bring that modularity back, and I believe Shane is intending to separate the part that operates on the number (like rounding, tailoring etc.) into `FixedDecimal` and similar helper structs and effectively removing the benefit of clustering all types of numerical operations in a single formatter.
@sffc can you share your take on this if I misunderstood you?

if that hypothesis is correct we have a way to get modular and lean `CurrencyFormatter`, `MeasureUnitFormatter`, `RuleBasedNumberFormatter`, `RelativeTimeFormatter`, `DurationFormatter` and so on into their own components and keep each one small without paying with an overhead when all of them are in use.

This issue has been filled to discuss that and verify if we're all on the same page about how we want to tackle that topic.

**Comment by sffc** [comment]
I am of the opinion that crates are not the most effective way to go about modularization.  I have written in wrapper-layer.md that I believe we can use dead code elimination to achieve modularization in a much more effective way.

**Comment by sffc** [comment]
> I believe Shane is intending to separate the part that operates on the number (like rounding, tailoring etc.) into FixedDecimal and similar helper structs and effectively removing the benefit of clustering all types of numerical operations in a single formatter.
> @sffc can you share your take on this if I misunderstood you?

FixedDecimal is intended as a type that preserves leading and trailing zeros on input and output of NumberFormat, which is an important feature we largely lack in 402.  Rounding operations cannot be split from NumberFormat because rounding depends on locale data for currencies, compact decimals, and measurement units.

**Comment by sffc** [comment]
2020-12-04 discussion:

- We all agree that the FFI (logical) API layer should be as modular as possible. So this question is only about the ergonomic layer.
- @mihnita and @sffc agree that it is useful to have a super-lightweight class that maps from FixedDecimal to localized string, with no currencies, compact notation, measurement units, etc.  This would be used in DateTimeFormat to reduce the dependency weight.
- @sffc points out that there is no clean way to perform inheritance between the different types of higher-level formatters.  For example, CurrencyFormatter is one code path, CompactFormatter is another, and CompactCurrencyFormatter is yet a third.  The three do not encapsulate each other.
- @nciric points out that we can start with the two ergonomic classes (simple and kitchen sink), and create modular classes as needed to help platforms perform dead code elimination
- @zbraniecki agrees; we can extract the second tier of formatters when needed

**Comment by sffc** [comment]
More specifically, here is how I see the breakdown of features going into FixedDecimalFormat (lower level) versus KitchenSinkNumberFormat (higher level):

### FixedDecimalFormat

**What:** Pass-through formatter for FixedDecimal, applying localized symbols but no arithmetic.

**Features:**

- Localized decimal digits
- Grouping separators
- Sign display
- Scientific notation\*

*\* Depends on the chosen design of #228*

### KitchenSinkNumberFormat

**What:** A larger, data-driven formatter supporting a larger set of UTS 35

**Features:**

- Currency\*
- Measurement units
- Compact notation
- Percentages

*\* "Currency" encompasses currency spacing rules, currency rounding, symbol resolution, etc.*

### Note on Rounding

Rounding is a big chunk of the logic in ICU NumberFormatter.  Unfortunately, it needs to be coupled with at least KitchenSinkNumberFormat, because the algorithm for selecting a compact form and applying a currency both require rounding the number based on locale data.

**Comment by younies** [comment]
is this fixed ?

**Comment by sffc** [comment]
> is this fixed ?

No, this is the meta-issue tracking the full Intl.NumberFormat implementation, including everything for currencies, units, compact decimals, scientific, ...
